Add aggregates doc page (#559)
leancodepl/corelibrary: docs/domain/03_aggregates.md

Reviewer: I believe this would be benefitial to reference https://github.com/leancodepl/corelibrary/blob/v8.0-preview/docs/guides/01_creating_an_aggregate.md here

# Aggregates

Aggregates are part of DDD. To create one you have to create entity class inheriting [`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) or [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs).

> ## [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs)

Reviewer 2: Let's start with `IAggregateRoot`. Then, describe the `W/oOC` version separately as a diff to `IAggregateRoot`, not the other way around.

Optimistic Concurrency should be the default here.

In order to create aggregate root you should create class inheriting it. [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) requires you to give Id type.

Consider the following aggregate.

```csharp
public class User : IAggregateRootWithoutOptimisticConcurrency<Id<User>>
{

>     public Id<User> Id { get; private init; }

Reviewer: This type of Id is obsolete. Please switch to source generated ids

    public string Name { get; private init; } = null!;

    private User()
    { }

    public User(Id<User> id, string name)
    {
        Id = id;
        Name = name;
    }
}
```

### Parameterless constructor

Parameterless constructor is required by entity framework. It is private to not be accessible outside.

### Id

> [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) requires you to give Id type. Id property of your aggregate root has this type.

Reviewer: ```suggestion
[`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) requires you to specify identity type. Every aggregate has an `Id` property of the specified type.
```

## [`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs)

[`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) extends [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) by adding `DateTime` optimistic concurrency token.

Consider aggregate from previous example, but with optimistic concurrency.

```csharp
public class User : IAggregateRoot<Id<User>>
{
    public Id<User> Id { get; private init; }
    public string Name { get; private init; } = null!;

    DateTime IOptimisticConcurrency.DateModified { get; set; }

    private User()
    { }

    public User(Id<User> id, string name)
    {
        Id = id;
        Name = name;
    }
}
```

### `IOptimisticConcurrency.DateModified`

> `DateModified` is optimistic concurrency token managed by application code. It is managed by [`EFRepository`](../../src/Domain/LeanCode.DomainModels.EF/EFRepository.cs) and you shouldn't do it by yourself, unless you know what you do. It is written this way, instead of `public DateTime DateModified { get; set; }`, to make it not accessible outside [`EFRepository`](../../src/Domain/LeanCode.DomainModels.EF/EFRepository.cs).

Reviewer: ```suggestion
`DateModified` is optimistic concurrency token managed by application code. It is managed by [`EFRepository`](../../src/Domain/LeanCode.DomainModels.EF/EFRepository.cs) and you shouldn't do it by yourself. It is written this way, instead of `public DateTime DateModified { get; set; }`, to make it not accessible outside [`EFRepository`](../../src/Domain/LeanCode.DomainModels.EF/EFRepository.cs).
```

# Aggregates

> Aggregates are part of DDD. To create one you have to create entity class inheriting [`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) or [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs).

Reviewer 2: ```suggestion
Aggregates are part of DDD. To create one you have to create a class inheriting [`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) or [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs).
```

## [`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs)

> In order to create aggregate root you should create class inheriting it. [`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) requires you to specify Id type.

Reviewer 2: ```suggestion
[`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) requires you to specify Id type.
```

The first sentence is slightly redundant - it is basically a rephrasing of the first paragraph.

Reviewer 2: I'm also not really sure that talking about the `Id` type here adds anything - let's talk about the `Id` itself (why we need that).

Consider the following aggregate.

```csharp
public class User : IAggregateRoot<UserId>
{

>     public UserId Id { get; private init; }

Reviewer 2: Let's also define the `UserId` (w/o explanation).

    public string Name { get; private init; } = null!;

    DateTime IOptimisticConcurrency.DateModified { get; set; }

    private User()
    { }

    public User(UserId id, string name)
    {
        Id = id;
        Name = name;
    }
}
```

### Parameterless constructor

> Parameterless constructor is required by entity framework. It is private to not be accessible outside.

Reviewer 2: ```suggestion
Parameterless constructor is required by EntityFramework. It is private because it should not be used in other scenarios, especially application code.
```

### Id

> [`IAggregateRoot`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) requires you to specify identity type. Every aggregate has an `Id` property of the specified type.

Reviewer 2: Let's state that we recommend using source-generated IDs here.

### `IOptimisticConcurrency.DateModified`

`DateModified` is optimistic concurrency token managed by application code. It is managed by [`EFRepository`](../../src/Domain/LeanCode.DomainModels.EF/EFRepository.cs) and you shouldn't do it by yourself. It is written this way, instead of `public DateTime DateModified { get; set; }`, to make it not accessible outside [`EFRepository`](../../src/Domain/LeanCode.DomainModels.EF/EFRepository.cs).

## [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs)

> Id you don't want to use optimistic concurrency for your aggregate, you can use [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) instead.

Reviewer 2: ```suggestion
If you don't want to use optimistic concurrency for your aggregate, you can use [`IAggregateRootWithoutOptimisticConcurrency`](../../src/Domain/LeanCode.DomainModels/Model/IAggregateRoot.cs) instead.
```
